README.md
<!--

 (C) Copyright 2016-2017 Fluenda.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

-->

# ParCEFone [![Build Status](https://travis-ci.org/fluenda/ParCEFone.svg?branch=master)](https://travis-ci.org/fluenda/ParCEFone)
ParCEFone is a simple to use Common Event Format - CEF - log parser for the JVM.

It is written in JAVA and can be used by most JVM languages.

## Using it

If using Maven, add the following dependency:

```
<dependency>
    <groupId>com.fluenda</groupId>
    <artifactId>ParCEFone</artifactId>
    <version>1.2.4</version>
</dependency>
```

ParCEFone is licensed under ASL 2


CHANGELOG.md
<!--

 (C) Copyright 2016 Fluenda.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

-->

# 1.2.5
## Improvements
* [PERFORMANCE] Allow the reuse of Bean Validator
* Reduce dependencies 

# 1.2.3 & 1.2.4
## Bug Fixes
* Fix typos on CEF extension keys

# 1.2.2
## Bug Fixes
* Adjusts timeRegex match to match non-ASCII characters

# 1.2.1
## Improvements
* Allow developers to control the locale used when parsing Date 
  extensions (defaulting to Locale.ENGLISH for backward compatibility)
* Improve documentation
## Bug Fixes
* ParCEFone Data parsing was locale dependent

# 1.2.0
## Improvements
* Dependency version bump and adoption of semantic versioning (no further 
  beaking changes to be introduced in minor versions)

# 1.1.1
## Bug Fixes
* Fixes IndexOutOfBoundsException when Parser is fed with totally bogus 
  input

# 1.1 
## Improvements
* Cover all 9 possible timestamp formats decribed in Appendix A

## Major changes
* All date related fields are now Date objects (breaking)
* getExtensions has been renamed to getExtension (breaking)

## Bug Fixes
* Properly handle float and double fields

# 1.0

The wild west...


src/main/java/com/fluenda/parcefone/parser/CEFParser.java
/*
 * (C) Copyright 2016 Fluenda.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.fluenda.parcefone.parser;

import com.fluenda.parcefone.event.CEFHandlingException;
import com.fluenda.parcefone.event.CefRev23;
import com.fluenda.parcefone.event.CommonEvent;

import javax.validation.ConstraintViolation;
import javax.validation.Validation;
import javax.validation.Validator;
import java.nio.charset.Charset;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A Common Event Format (CEF) parser used to convert String or byte array into a Map containing the <b>parsed and
 * validated</b> CEF fields
 * The parser does not make any assertion in regards to thread safety. Proceed with care.
 */
public class CEFParser {
    final static Logger logger = LoggerFactory.getLogger(CEFParser.class);
    Validator validator;


    /**
    *  Creates a CEFParser instance utilizing the default Bean Validator.
     */
    public CEFParser() {

    }

    /**
     *  Creates a CEFParser instance utilizing thread-safe Beans Validator. The use of this constructor should result in significantly higher
     *  throughput when performing multiple instatiations of CEFParser.
     * @param validator A JSR-303 complianceValidator such as Hibernate or Apache bVal
     */
    public CEFParser(Validator validator) {
        this.validator = validator;
    }

    /**
     * @return CommonEvent
     * @param cefByteArray byte [] containing the CEF message to be parsed - Array will be converted String using UTF-8
     */
    public CommonEvent parse(byte [] cefByteArray)  {
        String cefString;
        cefString = new String(cefByteArray, Charset.forName("UTF-8"));
        return this.parse(cefString, false);
    }

    /**
     * @return CommonEvent
     * @param cefByteArray byte [] containing the CEF message to be parsed - Array will be converted String using UTF-8
     * @param validate Boolean if parser should validate values beyond type compatibility (e.g. Values within acceptable lengths, value lists, etc)
     */
    public CommonEvent parse(byte [] cefByteArray, boolean validate)  {
        String cefString;
        cefString = new String(cefByteArray, Charset.forName("UTF-8"));
        return this.parse(cefString, validate);
    }

    /**
     * @return CommonEvent
     * @param cefByteArray byte [] containing the CEF message to be parsed - Array will be converted String using UTF-8
     * @param validate Boolean if parser should validate values beyond type compatibility (e.g. Values within acceptable lengths, value lists, etc)
     * @param locale The locale to be used when parsing dates (so that parser can handle both jul (en_US) and juil.(fr_FR)
     */
    public CommonEvent parse(byte [] cefByteArray, boolean validate, Locale locale)  {
        String cefString;
        cefString = new String(cefByteArray, Charset.forName("UTF-8"));
        return this.parse(cefString, validate, locale);
    }


    /**
     * <p>
     * Converts a CEF formatted String into a {@link CommonEvent} object without enforcing strict validation.
     * <p>
     * The use of this method is discouraged and future versions may deprecate its use.
     * @param cefString String containing the CEF message to be parsed
     * @return CommonEvent
     */
    public CommonEvent parse(String cefString)  {
        return this.parse(cefString, false);
    }

    /**
     * <p>
     * Converts a CEF formatted String into a {@link CommonEvent} object with exposed control over strict validation.
     * <p>
     * All CEF extension fields containing Dates are processed with the {@link Locale Locale.ENGLISH}.
     * @param cefString String containing the CEF message to be parsed
     * @param validate Boolean if parser should validate values beyond type compatibility (e.g. Values within acceptable lengths, value lists, etc)
     * @return CommonEvent
     */
    public CommonEvent parse(String cefString, final boolean validate)  {
        return this.parse(cefString, validate, Locale.ENGLISH);
    }

    /**
     * Converts a CEF formatted String into a {@link CommonEvent} object with exposed control over validation and the
     * {@link Locale} used to parse fields containing {@link Date Dates}
     * @param cefString String containing the CEF message to be parsed
     * @param validate Boolean if parser should validate values beyond type compatibility (e.g. Values within acceptable lengths, value lists, etc)
     * @param locale The locale to be used when parsing dates (so that parser can handle both jul (en_US) and juil.(fr_FR)
     * @return CommonEvent
     */
    public CommonEvent parse(String cefString, final boolean validate, Locale locale)  {

        int cefHeaderSize = 7;
        CommonEvent cefEvent = new CefRev23(locale);

        // Note how split number of splits is cefHeaderSize + 1. This is because the final split
        // should be the body of the CEF message
        // Compiled pattern is equivalent to "(?<!\\\\)" + Pattern.quote("|")
        final String[] extractedMessage = cefString.split("(?<!\\\\)" + Pattern.quote("|"), cefHeaderSize + 1);

        // CEF header misses values
        if (extractedMessage.length < cefHeaderSize) {
            if (logger.isDebugEnabled()) {
                logger.debug("CEF message failed validation");
            }
            return null;
        }

        final HashMap<String, Object> headers = new HashMap<String, Object>();
        headers.put("version", Integer.valueOf(extractedMessage[0].substring(extractedMessage[0].length() - 1)));
        headers.put("deviceVendor", extractedMessage[1]);
        headers.put("deviceProduct", extractedMessage[2]);
        headers.put("deviceVersion", extractedMessage[3]);
        headers.put("deviceEventClassId", extractedMessage[4]);
        headers.put("name", extractedMessage[5]);
        headers.put("severity", extractedMessage[6]);

        final HashMap<String, String> extensions = new HashMap<String, String>();


        final String ext = extractedMessage[7];

        // Compiled pattern is equivalent to String extensionRegex = "(?<!\\\\)" + Pattern.quote("=");
        final Matcher matcher = Pattern.compile("(?<!\\\\)" + Pattern.quote("=")).matcher(ext);
        final Matcher valueMatcher = Pattern.compile("(?<!\\\\)" + Pattern.quote("=")).matcher(ext);
        int index = 0;

        while (matcher.find()) {
            String key = ext.substring(index, matcher.end() - 1);

            // Capture the start of the value (first char after delimiter match);
            int valueStart = matcher.end();

            // Handle all but last extension
            if (valueMatcher.find(valueStart)) {
                // FInd the next match to determine the maximum length of the value
                int nextMatch = valueMatcher.start();
                // Find the last space prior to next match (i.e. last char before next key)
                int lastSpace = ext.lastIndexOf(" ", nextMatch);

                // Copy the value between the value start (i.e. this match)
                // and the lastSpace (i.e. last char prior to next key)
                String value = ext.substring(valueStart, lastSpace);

                // Put to map.
                extensions.put(key, value);

                // Update index to the last character before the next key
                index = lastSpace + 1;

                // Treat the last KV (if match is true)
            } else if (valueMatcher.find(valueStart - 1)) {
                // We are handling the final character, value end if newline at the
                // end of string
                int valueEnd = ext.length();

                String value = ext.substring(valueStart, valueEnd);
                extensions.put(key, value);
                // Update the index to the end of the string so no matches are possible
                index = valueEnd;
            }
        }

        try {
            cefEvent.setHeader(headers);
            cefEvent.setExtension(extensions);

        } catch (CEFHandlingException e) {
            logger.error(e.toString());
            return null;
        }

        if (validate) {
            if (validator == null) {
                // Since the validator wasn't initiated previously, create a new one;
                this.validator = Validation.buildDefaultValidatorFactory().getValidator();;
            }

            Set<ConstraintViolation<CommonEvent>> validationResult = validator.validate(cefEvent);

            if (validationResult.size() > 0) {
            if (logger.isDebugEnabled()) {
                logger.debug("CEF message failed validation");
            }
                return null;
            } else {
                return cefEvent;
            }
        } else {
            return cefEvent;
        }
    }
}



src/main/java/com/fluenda/parcefone/event/CommonEvent.java
/*
 * (C) Copyright 2016 Fluenda.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.fluenda.parcefone.event;

import java.util.Map;

public abstract class CommonEvent {
    // Implements a " struct like"  class that implements the Common Event
    // Format v23 as described here:
    // https://www.protect724.hpe.com/servlet/JiveServlet/downloadBody/1072-102-9-20354/CommonEventFormatv23.pdf

    // Define getters and seters
    public abstract void setHeader(Map<String, Object> headers) throws CEFHandlingException;

    public abstract void setExtension(Map<String, String> extensions) throws CEFHandlingException;

    public abstract Map<String, Object> getHeader() throws CEFHandlingException;

    public abstract Map<String, Object> getExtension(boolean populatedOnly) throws CEFHandlingException;
}


src/main/java/com/fluenda/parcefone/event/CEFHandlingException.java
/*
 * (C) Copyright 2016 Fluenda.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.fluenda.parcefone.event;

public class CEFHandlingException extends Exception {
    public CEFHandlingException() {

    }
    public CEFHandlingException(String message) {
        super(message);
    }
    public CEFHandlingException(String message, Throwable cause) {
        super(message,cause);
    }
}


src/main/java/com/fluenda/parcefone/event/CefRev23.java
/*
 * (C) Copyright 2016 Fluenda.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.fluenda.parcefone.event;

import com.martiansoftware.macnificent.MacAddress;


import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;
import java.lang.reflect.Field;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;

/**
 * Implements the Common Event Format (CEF) as documented by
 * <a href="https://www.protect724.hpe.com/servlet/JiveServlet/downloadBody/1072-102-9-20354/CommonEventFormatv23.pdf">
 *     revision 23
 * </a>
 * (Retrieved on August 2016).
 *
 */
public class CefRev23 extends CommonEvent {

    // Note the conflict with javax.validation.constraints.Pattern...
    final private java.util.regex.Pattern timeRegex =  java.util.regex.Pattern.compile(
                    "(?<MONTH>\\S+(\\.)?)\\s(?<DAY>\\d{2})\\s(?:(?<YEAR>\\d{4})(?:\\s))?" +
                    "(?<HOUR>[012][0-9]):(?<MINUTE>[0-5][0-9]):(?<SECOND>[0-5][0-9])" +
                    "(?:\\.(?<MILLI>\\d{3}))?(?:\\s(?<TZ>\\w+))?");

    final private Class<?> objClass = this.getClass();
    final private Field[] fields = objClass.getDeclaredFields();
    private ArrayList<String> populatedExtensions = new ArrayList<String>();

    // Implements a " struct like"  class that implements the Common Event
    // Format v23 as described here:
    // https://www.protect724.hpe.com/servlet/JiveServlet/downloadBody/1072-102-9-20354/CommonEventFormatv23.pdf

    private int version;
    private String deviceVendor;
    private String deviceProduct;
    private String deviceVersion;
    // Device Event Class ID is defined as int or String (yay!) treating as
    // string
    private String deviceEventClassId;
    private String name;
    private String severity;

    // The extension field and its list of KVs

    @Size(max = 63)
    private String act;

    @Size(max = 31)
    private String app;

    private Inet6Address c6a1;

    @Size(max = 1023)
    private String c6a1Label;

    private Inet6Address c6a2;

    @Size(max = 1023)
    private String c6a2Label;

    private Inet6Address c6a3;

    @Size(max = 1023)
    private String c6a3Label;

    private Inet6Address c6a4;

    @Size(max = 1023)
    private String c6a4Label;

    private float cfp1;

    @Size(max = 1023)
    private String cfp1Label;

    private float cfp2;

    @Size(max = 1023)
    private String cfp2Label;

    private float cfp3;

    @Size(max = 1023)
    private String cfp3Label;

    private float cfp4;

    @Size(max = 1023)
    private String cfp4Label;

    private long cn1;

    @Size(max = 1023)
    private String cn1Label;

    private long cn2;

    @Size(max = 1023)
    private String cn2Label;

    private long cn3;

    @Size(max = 1023)
    private String cn3Label;

    private long cnt;

    @Size(max = 4000)
    private String cs1;

    @Size(max = 1023)
    private String cs1Label;

    @Size(max = 4000)
    private String cs2;

    @Size(max = 1023)
    private String cs2Label;

    @Size(max = 4000)
    private String cs3;

    @Size(max = 1023)
    private String cs3Label;

    @Size(max = 4000)
    private String cs4;

    @Size(max = 1023)
    private String cs4Label;

    @Size(max = 4000)
    private String cs5;

    @Size(max = 1023)
    private String cs5Label;

    @Size(max = 4000)
    private String cs6;

    @Size(max = 1023)
    private String cs6Label;

    @Size(max = 255)
    private String destinationDnsDomain;

    @Size(max = 1023)
    private String destinationServiceName;

    private Inet4Address destinationTranslatedAddress;

    private int destinationTranslatedPort;

    private Date deviceCustomDate1;

    @Size(max = 1023)
    private String deviceCustomDate1Label;

    private Date deviceCustomDate2;

    @Size(max = 1023)
    private String deviceCustomDate2Label;

    // OMG! Device direction is binary!!!
    // 0 = inbound 1 is outbound
    @Min(0)
    @Max(1)
    private int deviceDirection;

    @Size(max = 255)
    private String deviceDnsDomain;

    @Size(max = 255)
    private String deviceExternalId;

    @Size(max = 1023)
    private String deviceFacility;

    @Size(max = 128)
    private String deviceInboundInterface;

    @Size(max = 255)
    private String deviceNtDomain;

    @Size(max = 128)
    private String deviceOutboundInterface;

    @Size(max = 128)
    private String devicePayloadId;

    @Size(max = 1023)
    private String deviceProcessName;

    private Inet4Address deviceTranslatedAddress;

    @Size(max = 1023)
    private String dhost;

    private MacAddress dmac;

    @Size(max = 255)
    private String dntdom;

    private int dpid;

    @Size(max = 1023)
    private String dpriv;

    @Size(max = 1023)
    private String dproc;

    @Max(65535)
    private int dpt;

    private Inet4Address dst;

    @Size(max = 255)
    private String dtz;

    @Size(max = 1023)
    private String duid;

    @Size(max = 1023)
    private String duser;

    private Inet4Address dvc;

    @Size(max = 100)
    private String dvchost;

    private MacAddress dvcmac;

    private int dvcpid;

    private Date end;

    @Size(max = 40)
    private String externalId;

    private Date fileCreateTime;

    @Size(max = 255)
    private String fileHash;

    @Size(max = 1023)
    private String field;

    private Date fileModificationTime;

    @Size(max = 1023)
    private String filePath;

    @Size(max = 1023)
    private String filePermission;

    @Size(max = 1023)
    private String fileType;

    private Date flexDate1;

    @Size(max = 128)
    private String flexDate1Label;

    private long flexNumber1;

    @Size(max = 128)
    private String flexNumber1Label;

    private long flexNumber2;

    @Size(max = 128)
    private String flexNumber2Label;

    @Size(max = 1023)
    private String flexString1;

    @Size(max = 128)
    private String flexString1Label;

    @Size(max = 1023)
    private String flexString2;

    @Size(max = 128)
    private String flexString2Label;

    @Size(max = 1023)
    private String fname;

    private int fsize;

    private int in;

    @Size(max = 1023)
    private String msg;

    private Date oldFileCreateTime;

    @Size(max = 255)
    private String oldFileHash;

    @Size(max = 1023)
    private String oldField;

    private Date oldFileModificationTime;

    @Size(max = 1023)
    private String oldFileName;

    @Size(max = 1023)
    private String oldFilePath;

    @Size(max = 1023)
    private String oldFilePermission;

    private int oldFileSize;

    @Size(max = 1023)
    private String oldFileType;

    private int out;

    @Size(max = 63)
    private String outcome;

    @Pattern(regexp = "tcp|udp", flags = Pattern.Flag.CASE_INSENSITIVE)
    @Size(max = 31)
    private String proto;

    @Size(max = 1023)
    private String reason;

    @Size(max = 1023)
    private String request;

    @Size(max = 1023)
    private String requestClientApplication;

    @Size(max = 2048)
    private String requestContext;

    @Size(max = 1023)
    private String requestCookies;

    @Size(max = 1023)
    private String requestMethod;

    private Date rt;

    @Size(max = 1023)
    private String shost;

    private MacAddress smac;

    @Size(max = 255)
    private String sntdom;

    @Size(max = 255)
    private String sourceDnsDomain;

    @Size(max = 1023)
    private String sourceServiceName;

    private Inet4Address sourceTranslatedAddress;

    private int sourceTranslatedPort;

    private int spid;

    @Size(max = 1023)
    private String spriv;

    @Size(max = 1023)
    private String sproc;

    @Max(65535)
    private int spt;

    private Inet4Address src;

    private Date start;

    @Size(max = 1023)
    private String suid;

    @Size(max = 1023)
    private String suser;

    @Min(0)
    @Max(3)
    private int type;

    @Size(max = 255)
    private String agentDnsDomain;

    @Size(max = 255)
    private String agentNtDomain;

    private Inet4Address agentTranslatedAddress;

    @Size(max = 200)
    private String agentTranslatedZoneExternalID;

    @Size(max = 2048)
    private String agentTranslatedZoneURI;

    @Size(max = 200)
    private String agentZoneExternalID;

    @Size(max = 2048)
    private String agentZoneURI;

    private InetAddress agt;

    @Size(max = 1023)
    private String ahost;

    @Size(max = 40)
    private String aid;

    private MacAddress amac;

    private Date art;

    @Size(max = 63)
    private String at;

    @Size(max = 255)
    private String atz;

    @Size(max = 31)
    private String av;

    @Size(max = 1023)
    private String cat;

    @Size(max = 200)
    private String customerExternalID;

    @Size(max = 2048)
    private String customerURI;

    @Size(max = 200)
    private String destinationTranslatedZoneExternalID;

    @Size(max = 2048)
    private String destinationTranslatedZoneURI;

    @Size(max = 200)
    private String destinationZoneExternalID;

    @Size(max = 2048)
    private String destinationZoneURI;

    @Size(max = 200)
    private String deviceTranslatedZoneExternalID;

    @Size(max = 2048)
    private String deviceTranslatedZoneURI;

    @Size(max = 200)
    private String deviceZoneExternalID;

    @Size(max = 2048)
    private String deviceZoneURI;

    private double dlat;

    private double dlong;

    private long eventId;

    @Size(max = 4000)
    private String rawEvent;

    private double slat;

    private double slong;

    @Size(max = 200)
    private String sourceTranslatedZoneExternalID;

    @Size(max = 2048)
    private String sourceTranslatedZoneURI;

    @Size(max = 200)
    private String sourceZoneExternalID;

    @Size(max = 2048)
    private String sourceZoneURI;

    private Locale dateLocale;


    public CefRev23(Locale locale) {
        super();
        this.dateLocale = locale;
    }

    public CefRev23() {
        super();

        // Defaults to ENGLISH locale when processing dates
        this.dateLocale = Locale.ENGLISH;
    }

    /**
    * @param headers A map containing the  keys and values of headers of CEF event
    * @throws CEFHandlingException when it has issues writing the values of the headers
    */
    public void setHeader(Map<String, Object> headers)  throws CEFHandlingException {
        for (String key : headers.keySet()) {
            try {
                Field field = objClass.getDeclaredField(key);
                Object value = headers.get(key);
                field.set(this, value);
            } catch (NoSuchFieldException | IllegalAccessException e) {
                throw new CEFHandlingException("Error writing values to headers", e);
            }
        }
    }

    /**
     * @return A map containing the keys and values of headers
     * @throws CEFHandlingException when it has issues reading the headers of CEF event
     */
    public Map<String, Object> getHeader() throws CEFHandlingException {
        final HashMap<String, Object> headers = new HashMap<String, Object>();
        List headersKeys = Arrays.asList(new String[] {"version", "deviceVendor", "deviceProduct", "deviceVersion", "deviceEventClassId", "name", "severity"});

        for (Field f: fields) {
            if (headersKeys.contains(f.getName())) {
                try {
                    headers.put(f.getName(), f.get(this));
                } catch (IllegalAccessException e) {
                    throw new CEFHandlingException("Error harvesting headers, e");
                }
            }
        }
        return headers;
    }


    /**
     * @param extensions A map containing the keys and values of extensions of CEF event
     * @throws CEFHandlingException when it has issues populating the extensions
     */
    public void setExtension(Map<String, String> extensions) throws CEFHandlingException {
        for (String key : extensions.keySet()) {
            try {
                Field field = objClass.getDeclaredField(key);
                String value = extensions.get(key);
                // Treat each Classes in a particular fashion

                // Inet, Inet4 and Inet6 address
                if (field.getType().equals(InetAddress.class) || field.getType().equals(Inet4Address.class) || field.getType().equals(Inet6Address.class)) {
                    try {
                        InetAddress inetAddress = InetAddress.getByName((String) value);
                        field.set(this, inetAddress);
                    } catch (UnknownHostException e) {
                        throw new CEFHandlingException("Error setting value to field " + key, e);
                    }

                // Date (timestamps) - Note we force a particular date format (set as private dateFormat above
                } else if (field.getType().equals(Date.class)) {
                    try {
                        // Use a ": "to match epoch vs. Dateformat
                        if (!value.toString().contains(":")) {
                            // This is epoch
                            field.set(this, new Date(Long.valueOf(value)));
                        } else {
                            // This is one of the remaining 8 possible values, regex it out...
                            Matcher matcher = timeRegex.matcher(value);

                            if (matcher.matches()) {
                                String year = matcher.group("YEAR") == null ? String.valueOf(Calendar.getInstance().get(Calendar.YEAR)) : matcher.group("YEAR") ;
                                String milli = matcher.group("MILLI") == null ? "000" : matcher.group("MILLI");

                                String regexDate =
                                        year + "-" +
                                        matcher.group("MONTH") + "-" +
                                        matcher.group("DAY") + " " +
                                        matcher.group("HOUR") + ":" +
                                        matcher.group("MINUTE") + ":" +
                                        matcher.group("SECOND") + "." +
                                        milli;
                                if (matcher.group("TZ") == null ) {
                                    field.set(this, dateFormat(false).parse(regexDate));
                                } else {
                                    regexDate = regexDate + " " + matcher.group("TZ");
                                    field.set(this, dateFormat(true).parse(regexDate));
                                }
                            }
                        }
                    } catch (ParseException|NumberFormatException e) {
                        throw new CEFHandlingException("Error setting value to field " + key, e);
                    }

                // Mac Addresses
                } else if (field.getType().equals(MacAddress.class)) {
                    field.set(this, new MacAddress(value));

                // Integers
                } else if (field.getType().equals(int.class)) {
                    field.set(this, Integer.valueOf(value));

                // Longs
                } else if (field.getType().equals(long.class)){
                    field.set(this, Long.valueOf(value));

                // Doubles
                } else if (field.getType().equals(double.class)) {
                    field.set(this, Double.valueOf(value));

                // Floats
                } else if (field.getType().equals(float.class)) {
                    field.set(this, Float.valueOf(value));

                // The rest (to be removed)
                } else {
                    field.set(this, value );
                }

                // Add the key to the populate keys listt
                populatedExtensions.add(key);
            } catch (NoSuchFieldException e) {
                // Ignore illegal key names;
                continue;
            } catch (IllegalAccessException e) {
                throw new CEFHandlingException("Error while setting CEF extension values", e);
            }
        }
    }


    /**
     * @param populatedOnly Boolean defining if Map should include all fields supported by {@link com.fluenda.parcefone.event.CefRev23}
     * @return A map containing the keys and values of CEF extensions
     * @throws CEFHandlingException when it hits issues (e.g. IllegalAccessException) reading the extensions
     */
    public Map<String, Object> getExtension(boolean populatedOnly) throws CEFHandlingException {

        final HashMap<String, Object> extensions = new HashMap<String, Object>();
        List headersKeys = Arrays.asList(new String[] {"version", "deviceVendor", "deviceProduct", "deviceVersion", "deviceEventClassId", "name", "severity"});

        for (Field f: fields) {
            // Exclude header objects
            if (!headersKeys.contains(f.getName())) {
                // check if populatedOnly was requested

                    try {
                        Object value = f.get(this);
                       if (!populatedOnly){
                            extensions.put(f.getName(), value);
                        } else if (populatedOnly && populatedExtensions.contains(f.getName())) {
                           extensions.put(f.getName(), value);
                       }
                    } catch (IllegalAccessException e) {
                        throw new CEFHandlingException("Error while harvesting keys", e);
                    }
                }
            }
        return extensions;
    }

    private SimpleDateFormat dateFormat(boolean containsTZ) {

        if (containsTZ) {
            return new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss.SSS zzz", this.dateLocale);
        } else {
            return new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss.SSS", this.dateLocale);
        }
    }



}
